### ISSUE TITLE
[Docker Hub] Image retention policies

### ISSUE THREAD

**Issue by pkennedyr** [docker/roadmap#17]
**Tell us about your request**
User-specified image management configurations for cleaning up unnecessary or unwanted repository tags.

**Which service(s) is this request for?**
Docker Hub

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
As an account admin/owner, I would like flexibility and control over cleaning up unnecessary or unwanted repository tags within my Hub repositories, so that I can control which tags should get deleted and prevent specific tags from being deleted.

**Are you currently working around this issue?**
N/A

**Additional context**
N/A

**Attachments**
N/A

**Comment by pkennedyr** [comment]
Would automatic detection of images that haven't been pulled in < X months be useful? What other information would be helpful to see in determining the staleness or continued relevancy of images?

**Comment by ingshtrom** [comment]
For us, I think that is most important. I think for well organized teams, the grouping by tagging is relevant. e.g. how many tags have a prefix of `test-`, I want those deleted weekly, while images that have a format of `v<semver>` are kept for a minimum of 30 days.

**Comment by mariusgrigaitis** [comment]
In relation to https://www.docker.com/blog/november-2024-updated-plans-announcement/ and storage limits (charges) being put in place from March 1st I hope that this feature gets into the product before then. 

There's currently no convenient tools to manage lifecycle policies for dockerhub available (that I was able to find) so it will make it really hard to manage storage without this.

**Comment by sheltoG** [comment]
@mariusgrigaitis yes! We'll continue enhancing Docker Hub with new features to manage images, including lifecycle policies, both leading up to March 1st and beyond.

**Comment by kohtala** [comment]
Hi.

We have a set of tags that are in production. We'd like to keep a copy of the production images to use them for vulnerability scanning. We can produce periodically a list of tags to keep and could upload it through the API. Preferably the API would be authenticated via OIDC. It seems the current immutable tags could define the images to keep from being deleted. [Update Immutable Tags](https://docs.docker.com/reference/api/hub/latest/#tag/repositories/operation/UpdateRepositoryImmutableTags) seems to take the list at once. (I see the documentation for [Verify Immutable Tags](https://docs.docker.com/reference/api/hub/latest/#tag/repositories/operation/VerifyRepositoryImmutableTags) is unclear what it verifies and the example response lists a tag that does not match the regexp in the request.) The Lifecycle Policy could also carry such list and have the list similarly updated at once.

In case our production list generation develops some bug and produces an empty list, we'd like to have some soft-deletion period of time the images are still recoverable. The soft-deleted tags still need to be unavailable for our CVE monitoring or other use that can alert us to the lost images. If soft-deleted images reappear in the policy to keep them, it would be helpful if the policy could have them become resurrected from soft-deletion automatically.

We share the same tags across a group of repositories. Any tag needs to be in all repositories or it's not needed in any. The application is not a single image, but runs from a group of images. An organization could have many groups with different tags. In other words, we not only tag the images, but the interface between the images. All the images with same tag originate from same monorepo version in source control.
